libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -537,26 +542,55 @@ std::optional<PbftRound> VoteManager::determineNewRound(PbftPeriod current_pbft_
 }
 
 PbftPeriod VoteManager::getRewardVotesPbftBlockPeriod() {
-  std::shared_lock lock(reward_votes_info_mutex_);
+  std::shared_lock lock(verified_votes_access_);
   return reward_votes_period_;
 }
 
-void VoteManager::resetRewardVotesInfo(PbftPeriod period, PbftRound round, const blk_hash_t& block_hash) {
-  {
-    std::scoped_lock lock(reward_votes_info_mutex_);
-
-    reward_votes_block_hash_ = block_hash;
-    reward_votes_period_ = period;
-    reward_votes_round_ = round;
+void VoteManager::resetRewardVotes(PbftPeriod period, PbftRound round, PbftStep step, const blk_hash_t& block_hash,
+                                   DbStorage::Batch& batch) {
+  // Save 2t+1 cert votes to database, remove old reward votes
+  std::scoped_lock lock(verified_votes_access_);
+  reward_votes_block_hash_ = block_hash;
+  reward_votes_period_ = period;
+  reward_votes_round_ = round;
+
+  auto found_period_it = verified_votes_.find(period);
+  // Add period
+  if (found_period_it != verified_votes_.end()) {
+    auto found_round_it = found_period_it->second.find(round);
+    // Add round
+    if (found_round_it != found_period_it->second.end()) {
+      auto found_step_it = found_round_it->second.step_votes.find(step);
+      // Add step
+      if (found_step_it != found_round_it->second.step_votes.end()) {
+        auto found_two_t_plus_one_voted_block =
+            found_round_it->second.two_t_plus_one_voted_blocks_.find(TwoTPlusOneVotedBlockType::CertVotedBlock);
+
+        // 2t+1 votes block already set
+        if (found_two_t_plus_one_voted_block != found_round_it->second.two_t_plus_one_voted_blocks_.end()) {
+          assert(found_two_t_plus_one_voted_block->second.first == block_hash);
+          auto found_voted_value_it = found_step_it->second.votes.find(block_hash);
+          // Add voted value
+          if (found_voted_value_it != found_step_it->second.votes.end()) {
+            std::vector<std::shared_ptr<Vote>> votes;
+            votes.reserve(found_voted_value_it->second.second.size());
+            for (const auto& tmp_vote : found_voted_value_it->second.second) {
+              votes.push_back(tmp_vote.second);
+            }
+            db_->replaceTwoTPlusOneVotesToBatch(TwoTPlusOneVotedBlockType::CertVotedBlock, votes, batch);
+            db_->removeRewardVotes(reward_votes_, batch);
+            reward_votes_.clear();
+          }
+        }
+      }
+    }
   }
 
   LOG(log_dg_) << "Reward votes info reset to: block_hash: " << block_hash << ", period: " << period
                << ", round: " << round;
 }
 
 bool VoteManager::isValidRewardVote(const std::shared_ptr<Vote>& vote) const {
-  std::shared_lock lock(reward_votes_info_mutex_);
```

Comment:
so what is protecting reward_votes_block_hash_ and reward_votes_period_ now ?

Reply:
I have put everything under the same mutex verified_votes_access_. Maybe the name of the mutex should be changed. This method isValidRewardVote is private and it is only called within this class from the code already protected by the mutex.